tools/autograd/templates/variable_factories.h

```diff
@@ -17,6 +18,125 @@
+    ListInitTensor(T scalar) :         \
+        scalar_(scalar), init_list_(), \
+        sizes_(),                      \
+        scalar_type_(at::k##S),        \
+        type_(ListInitTensorType::Scalar) {}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+    ListInitTensor(std::initializer_list<ListInitTensor> init_list) :
+        scalar_(),
+        init_list_(init_list),
+        sizes_(),
+        scalar_type_(init_list.begin()->scalar_type_),
+        type_(ListInitTensorType::InitList) {
+      const ListInitTensor& first_elem = *(init_list.begin());
+      for (const auto& elem : init_list) {
+        TORCH_CHECK(elem.scalar_type_ == first_elem.scalar_type_,
+          "Expected all elements of the tensor to have the same scalar type: ",
+          first_elem.scalar_type_,
+          ", but got element of scalar type: ",
+          elem.scalar_type_);
+        TORCH_CHECK(elem.sizes_ == first_elem.sizes_,
+          "Expected all sub-tensors to have sizes: ",
+          first_elem.sizes_,
+          " (e.g. ", first_elem, "), ",
+          "but got sub-tensor ",
+          elem,
+          " with sizes: ",
+          elem.sizes_);
+      }
+      sizes_.push_back(init_list.size());
+      sizes_.insert(sizes_.end(), first_elem.sizes_.begin(), first_elem.sizes_.end());
+    }
+
+    const c10::Scalar& scalar() const {
+      return scalar_;
+    }
+
+    const std::initializer_list<ListInitTensor>& init_list() const {
+      return init_list_;
+    }
+
+    const std::vector<int64_t>& sizes() const {
+      return sizes_;
+    }
+
+    const c10::ScalarType& scalar_type() const {
+      return scalar_type_;
+    }
+
+    const ListInitTensorType& type() const {
+      return type_;
+    }
+
+    at::Tensor to_tensor(const at::TensorOptions& options) const {
+      at::Tensor tensor = ([&]() {
+        at::AutoNonVariableTypeMode non_var_type_mode(true);
+        return at::empty(sizes_, at::TensorOptions(options).is_variable(false));
+      })();
+      fill_tensor(tensor);
+      return tensor;
+    }
+
+    void pretty_print_recursive(std::ostream& stream) const {
+      if (type_ == ListInitTensorType::Scalar) {
+        AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Bool, scalar_type_, "ListInitTensor_pretty_print_scalar", [&] {
+          stream << const_cast<c10::Scalar&>(scalar_).to<scalar_t>();
+        });
+      } else if (type_ == ListInitTensorType::InitList) {
+        stream << "{";
+        for (const ListInitTensor* it = init_list_.begin(); it != init_list_.end(); it++) {
+          it->pretty_print_recursive(stream);
+          if (std::next(it) != init_list_.end()) stream << ", ";
+        }
+        stream << "}";
+      }
+    }
+
+   private:
+    void fill_tensor(at::Tensor tensor) const {
+      int index = 0;
+      for (const auto& elem : init_list_) {
+        if (elem.type_ == ListInitTensorType::Scalar) {
+          at::NoGradGuard guard;
+          tensor[index].fill_(elem.scalar());
```

Comment:
This seems dangerously slow, especially for CUDA tensors (filling each element would incur the 6 us CUDA launch overhead). Maybe for CUDA tensors it is better to fill on CPU first, then transfer to CUDA? It would be good to benchmark this to determine what the right solution is

Reply:
Thanks for the suggestion - I did a benchmark on it and fill-on-CPU-transfer-to-CUDA is significantly faster than fill-on-CUDA. Specifically, the benchmark I ran is `torch::tensor({{{{{{{{1.0, 2.0, 3.0}}}}}, {{{{{4.0, 5.0, 6.0}}}}}, {{{{{7.0, 8.0, 9.0}}}}}}}}, torch::kCUDA)`, and fill-on-CUDA takes about 101059 ns, while fill-on-CPU-transfer-to-CUDA takes about 74118 ns, so we are saving (101059 - 74118) / 9 ~= 3000 ns = 3 us per element, which is of the same order of magnitude of the CUDA launch overhead (that we cut down by launching only 1 CUDA kernel).